File: packages/draftfirst/src/smarttype.ts
/**
 * Draft First Screenwriting Engine vocabulary extraction.
 *
 * Derives character names, locations, scene times, transitions, and heading
 * prefixes from a screenplay without retaining state.
 */

import type { Screenplay } from './types.js';

/** Canonical scene-time tokens shared by parsing and prediction. */
export const SCENE_TIME_VALUES: readonly string[] = Object.freeze([
	'DAY',
	'NIGHT',
	'MORNING',
	'AFTERNOON',
	'EVENING',
	'DAWN',
	'DUSK',
	'SUNRISE',
	'SUNSET',
	'MAGIC HOUR',
	'MIDNIGHT',
	'LATER',
	'CONTINUOUS',
	'MOMENTS LATER',
	'SAME',
	'SAME TIME',
	'THE NEXT DAY',
	'DAYS LATER',
	'WEEKS LATER',
	'MONTHS LATER',
	'YEARS LATER',
	'FLASHBACK',
	'PRESENT DAY',
	'FUTURE'
]);

const SCENE_TIME_SET = new Set(SCENE_TIME_VALUES);

export interface SmartTypeData {
	/** Character names with extensions stripped — MOLLY, not MOLLY (V.O.). */
	characters: string[];
	/** Location parts of scene headings — POLICE STATION. */
	locations: string[];
	/** Time-of-day parts of scene headings — DAY, NIGHT, LATER. */
	times: string[];
	/** Transitions used — CUT TO:, SMASH CUT TO:. */
	transitions: string[];
	/** Scene heading prefixes seen — INT., EXT., INT./EXT. */
	prefixes: string[];
}

export interface SceneHeadingParts {
	prefix: string;
	location: string;
	time: string;
}

/** Strip cue extensions — (V.O.), (O.S.), (O.C.), (CONT'D), (SUBTITLE)… */
export function stripCueExtensions(cue: string): string {
	return cue
		.replace(/\s*\((?:V\.?O\.?|O\.?S\.?|O\.?C\.?|CONT['’]?D|SUBTITLE|PRE-?LAP|FILTERED|INTO (?:PHONE|RADIO|COMMS?)[^)]*)\)\s*/gi, '')
		.trim();
}

/** Split a scene heading into prefix / location / time parts.
   Parts are canonical UPPERCASE: a heading is a structural token, not
   prose — every consumer (memory, habits, drift checks) compares parts
   case-insensitively, and the model cannot be trusted to carry canonical
   case (blur without commit, paste, and mixed-case imports all leak). */
export function splitSceneHeading(heading: string): SceneHeadingParts {
	const m = /^(INT\.?\/EXT\.?|INT\/EXT|I\/E|INT|EXT|EST)[. ]?\s*/i.exec(heading.trim());
	const rawPrefix = m ? m[1].toUpperCase().replace(/\.$/, '') : '';
	const prefix = rawPrefix === '' ? '' : rawPrefix === 'I/E' ? 'I/E' : `${rawPrefix}.`;
	let rest = m ? heading.trim().slice(m[0].length) : heading.trim();

	let time = '';
	const parts = rest.split(/\s+[-–—]\s+/).map((part) => part.trim());
	if (parts.length > 1) {
		/* A heading may carry a modifier after time (`- DAY - ESTABLISHING`).
		   Select the first recognised time segment; locations may themselves
		   contain spaced dashes (`54TH ST - UPTOWN - DAY`). */
		let timeAt = parts.findIndex((part, index) => index > 0 && SCENE_TIME_SET.has(part.toUpperCase()));
		if (timeAt < 0) timeAt = parts.length - 1; // preserve custom times
		time = parts[timeAt];
		rest = parts.slice(0, timeAt).join(' - ');
	}
	return { prefix, location: rest.toUpperCase(), time: time.toUpperCase() };
}

function pushUnique(list: string[], seen: Set<string>, value: string): void {
	if (value !== '' && !seen.has(value)) {
		seen.add(value);
		list.push(value);
	}
}

/** Derive autocomplete/consistency vocabularies from the document. */
export function collectSmartType(script: Screenplay): SmartTypeData {
	const data: SmartTypeData = {
		characters: [],
		locations: [],
		times: [],
		transitions: [],
		prefixes: []
	};
	const seen = {
		characters: new Set<string>(),
		locations: new Set<string>(),
		times: new Set<string>(),
		transitions: new Set<string>(),
		prefixes: new Set<string>()
	};

	for (const el of script.elements) {
		if (el.type === 'character') {
			/* canonical case at the collection boundary — the engine never
			   trusts the model's casing (blur, paste, import all leak raw) */
			pushUnique(data.characters, seen.characters, stripCueExtensions(el.text).toUpperCase());
		} else if (el.type === 'scene') {
			const { prefix, location, time } = splitSceneHeading(el.text);
			pushUnique(data.prefixes, seen.prefixes, prefix);
			pushUnique(data.locations, seen.locations, location);
			pushUnique(data.times, seen.times, time);
		} else if (el.type === 'transition') {
			pushUnique(data.transitions, seen.transitions, el.text.trim().toUpperCase());
		}
	}
	return data;
}

/** Return location pairs that may differ only by articles or punctuation. */
export function findLocationDrift(script: Screenplay): Array<{ a: string; b: string }> {
	const { locations } = collectSmartType(script);
	const normalise = (s: string) =>
		s.replace(/^(THE|A|AN)\s+/i, '').replace(/['’.,-]/g, '').replace(/\s+/g, ' ').trim().toLowerCase();
	const entries = locations.map((surface, index) => ({ surface, index, key: normalise(surface) }));
	const postings = new Map<string, number[]>();
	for (const entry of entries) {
		for (const token of new Set(entry.key.split(' ').filter(Boolean))) {
			const list = postings.get(token);
			if (list) list.push(entry.index);
			else postings.set(token, [entry.index]);
		}
	}

	const found = new Set<string>();
	for (const entry of entries) {
		if (entry.key === '') continue;
		const tokens = entry.key.split(' ');
		let candidates: readonly number[] = [];
		for (const token of tokens) {
			const list = postings.get(token) ?? [];
			if (candidates.length === 0 || list.length < candidates.length) candidates = list;
		}
		for (const otherIndex of candidates) {
			if (otherIndex === entry.index) continue;
			const a = Math.min(entry.index, otherIndex);
			const b = Math.max(entry.index, otherIndex);
			const pairKey = `${a}:${b}`;
			if (found.has(pairKey)) continue;
			const left = entries[a].key;
			const right = entries[b].key;
			const paddedLeft = ` ${left} `;
			const paddedRight = ` ${right} `;
			if (
				left === right ||
				paddedLeft.includes(paddedRight) ||
				paddedRight.includes(paddedLeft)
			) {
				found.add(pairKey);
			}
		}
	}

	return [...found]
		.map((key) => key.split(':').map(Number) as [number, number])
		.sort(([a1, b1], [a2, b2]) => a1 - a2 || b1 - b2)
		.map(([a, b]) => ({ a: locations[a], b: locations[b] }));
}

File: apple/eDraftEngine/Sources/EDraftEngine/Predict/SmartType.swift
import Foundation

/// Vocabulary extraction — character names, locations, scene times,
/// transitions, and heading prefixes — ported from the TypeScript engine's
/// `smarttype.ts`. Pinned by the predict corpus (which exercises these
/// through every candidate generator).
public enum SmartType {

    /// Canonical scene-time tokens shared by parsing and prediction.
    public static let sceneTimeValues: [String] = [
        "DAY", "NIGHT", "MORNING", "AFTERNOON", "EVENING", "DAWN", "DUSK",
        "SUNRISE", "SUNSET", "MAGIC HOUR", "MIDNIGHT", "LATER", "CONTINUOUS",
        "MOMENTS LATER", "SAME", "SAME TIME", "THE NEXT DAY", "DAYS LATER",
        "WEEKS LATER", "MONTHS LATER", "YEARS LATER", "FLASHBACK",
        "PRESENT DAY", "FUTURE",
    ]

    private static let sceneTimeSet = Set(sceneTimeValues)

    public struct Data: Equatable, Sendable {
        /// Character names with extensions stripped — MOLLY, not MOLLY (V.O.).
        public var characters: [String] = []
        /// Location parts of scene headings — POLICE STATION.
        public var locations: [String] = []
        /// Time-of-day parts of scene headings — DAY, NIGHT, LATER.
        public var times: [String] = []
        /// Transitions used — CUT TO:, SMASH CUT TO:.
        public var transitions: [String] = []
        /// Scene heading prefixes seen — INT., EXT., INT./EXT.
        public var prefixes: [String] = []
    }

    public struct HeadingParts: Equatable, Sendable {
        public var prefix: String
        public var location: String
        public var time: String
    }

    /// Strip cue extensions — (V.O.), (O.S.), (O.C.), (CONT'D), (SUBTITLE)…
    /// Case-insensitive; accepts curly apostrophes, matching the JS regex.
    /// The one bare suffix the corpus witnesses is the transcript's V/O —
    /// "FRANK V/O" (pasted-26 ×299); O.S./O.C. stay parenthesised-only, the
    /// ambiguity the "MARA O.S." pin guards.
    public static func stripCueExtensions(_ cue: String) -> String {
        /* Every alternative sits inside a literal `\(...\)`, so a cue with
           no parenthesis and no V/O tail cannot match. Building the regex
           is the expensive part — it is not `Sendable`, so it cannot be
           shared — and the paginator asks this question of every dialogue
           block on every pass. Answer the common case without paying for
           it. */
        guard cue.contains("(") || cue.hasSuffix("V/O") else {
            return cue.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        var stripped = cue
        if cue.contains("(") {
            let pattern = #/(?i)\s*\((?:V\.?O\.?|O\.?S\.?|O\.?C\.?|CONT['’]?D|SUBTITLE|PRE-?LAP|FILTERED|INTO (?:PHONE|RADIO|COMMS?)[^)]*)\)\s*/#
            stripped = stripped.replacing(pattern, with: "")
        }
        // \s+V\/O$ — the bare transcript suffix
        if stripped.hasSuffix("V/O"),
           stripped.dropLast(3).last?.isWhitespace == true {
            stripped = String(stripped.dropLast(3))
        }
        return stripped.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    // MARK: - Scene heading splitting

    /// Intro tokens in JS-regex alternation order (compound forms first so
    /// `INT./EXT.` is never clipped to `INT.`, and the reversed hand six
    /// corpus files write — EXT/INT. — reads the same way). Dots are part
    /// of the token; a separate optional `[. ]` and `\s*` run follows.
    private static let introTokens = [
        "INT./EXT.", "INT./EXT", "INT/EXT.", "INT/EXT",
        "EXT./INT.", "EXT./INT", "EXT/INT.", "EXT/INT",
        "I/E", "INT", "EXT", "EST",
    ]

    /// Matches `/^(INT\.?\/EXT\.?|INT\/EXT|I\/E|INT|EXT|EST)[. ]?\s*/i` and
    /// returns the raw prefix token plus the total characters consumed.
    static func matchIntro(_ heading: String) -> (raw: String, consumed: Int)? {
        for token in introTokens where heading.range(of: token, options: [.caseInsensitive, .anchored]) != nil {
            var index = heading.index(heading.startIndex, offsetBy: token.count)
            if index < heading.endIndex && (heading[index] == "." || heading[index] == " ") {
                index = heading.index(after: index)
            }
            while index < heading.endIndex && heading[index].isJSWhitespace {
                index = heading.index(after: index)
            }
            return (token, heading.distance(from: heading.startIndex, to: index))
        }
        return nil
    }

    /// Splits on `/\s+[-–—]\s+/` — spaced dashes (hyphen, en, em).
    private static func splitOnSpacedDashes(_ text: String) -> [String] {
        var parts: [String] = []
        var cursor = text.startIndex
        var scan = text.startIndex
        while scan < text.endIndex {
            if "-–—".contains(text[scan]),
               scan > text.startIndex {
                /* require whitespace before the dash */
                let before = text.index(before: scan)
                guard text[before].isJSWhitespace else {
                    scan = text.index(after: scan)
                    continue
                }
                /* require whitespace after the dash */
                var after = text.index(after: scan)
                guard after < text.endIndex, text[after].isJSWhitespace else {
                    scan = text.index(after: scan)
                    continue
                }
                /* find the extent of the preceding whitespace run */
                var left = before
                while left > text.startIndex {
                    let earlier = text.index(before: left)
                    if !text[earlier].isJSWhitespace { break }
                    left = earlier
                }
                while after < text.endIndex && text[after].isJSWhitespace {
                    after = text.index(after: after)
                }
                /* JS `split` matches never overlap: when this dash's
                   preceding whitespace run was already consumed by the
                   previous separator, the dash is literal text, not a
                   separator. (`INT. LAB - - DAY` → ["INT. LAB", "- DAY"].)
                   Clamping instead would append "" and diverge from JS. */
                guard left >= cursor else {
                    scan = text.index(after: scan)
                    continue
                }
                parts.append(String(text[cursor..<left]))
                cursor = after
                scan = after
                continue
            }
            scan = text.index(after: scan)
        }
        parts.append(String(text[cursor...]))
        return parts
    }

    /// Split a scene heading into prefix / location / time parts. Parts are
    /// canonical UPPERCASE — the model cannot be trusted to carry canonical
    /// case (blur without commit, paste, and mixed-case imports all leak).
    public static func splitSceneHeading(_ heading: String) -> HeadingParts {
        let trimmed = heading.trimmingCharacters(in: .whitespacesAndNewlines)
        var rawPrefix = ""
        var rest = trimmed
        if let match = matchIntro(trimmed) {
            rawPrefix = match.raw.uppercased()
            if rawPrefix.hasSuffix(".") { rawPrefix.removeLast() }
            rest = String(trimmed.dropFirst(match.consumed))
        }
        let prefix = rawPrefix.isEmpty ? "" : rawPrefix == "I/E" ? "I/E" : rawPrefix + "."

        var time = ""
        let parts = splitOnSpacedDashes(rest).map {
            $0.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        var location = rest
        if parts.count > 1 {
            /* A heading may carry a modifier after time (`- DAY -
               ESTABLISHING`); locations may themselves contain spaced
               dashes (`54TH ST - UPTOWN - DAY`). */
            var timeAt = parts.indices.dropFirst().first(where: {
                sceneTimeSet.contains(parts[$0].uppercased())
            })
            if timeAt == nil { timeAt = parts.count - 1 }  // preserve custom times
            time = parts[timeAt!]
            location = parts[..<timeAt!].joined(separator: " - ")
        }
        return HeadingParts(
            prefix: prefix,
            location: location.uppercased(),
            time: time.uppercased()
        )
    }

    // MARK: - Collection

    private static func pushUnique(_ list: inout [String], _ seen: inout Set<String>, _ value: String) {
        if !value.isEmpty && !seen.contains(value) {
            seen.insert(value)
            list.append(value)
        }
    }

    /// Derive autocomplete/consistency vocabularies from the document.
    public static func collect(_ script: Screenplay) -> Data {
        var data = Data()
        var seenCharacters = Set<String>()
        var seenLocations = Set<String>()
        var seenTimes = Set<String>()
        var seenTransitions = Set<String>()
        var seenPrefixes = Set<String>()

        for element in script.elements {
            switch element.type {
            case .character:
                /* canonical case at the collection boundary */
                pushUnique(&data.characters, &seenCharacters,
                           stripCueExtensions(element.text).uppercased())
            case .scene:
                let parts = splitSceneHeading(element.text)
                pushUnique(&data.prefixes, &seenPrefixes, parts.prefix)
                pushUnique(&data.locations, &seenLocations, parts.location)
                pushUnique(&data.times, &seenTimes, parts.time)
            case .transition:
                pushUnique(&data.transitions, &seenTransitions,
                           element.text.trimmingCharacters(in: .whitespacesAndNewlines).uppercased())
            default:
                continue
            }
        }
        return data
    }
}
